File: wallet/src/de/schildbach/wallet/ui/HeaderBalanceFragment.java
/*
 * Copyright 2019 Dash Core Group
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.schildbach.wallet.ui;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import org.bitcoinj.core.Coin;
import org.bitcoinj.utils.Fiat;
import org.dash.wallet.common.Configuration;
import org.dash.wallet.common.ui.CurrencyTextView;
import org.dash.wallet.common.util.GenericUtils;

import javax.annotation.Nullable;

import dagger.hilt.android.AndroidEntryPoint;
import de.schildbach.wallet.AppDatabase;
import de.schildbach.wallet.Constants;
import de.schildbach.wallet.WalletApplication;
import de.schildbach.wallet.data.BlockchainState;
import de.schildbach.wallet.data.DashPayProfile;
import de.schildbach.wallet.rates.ExchangeRate;
import de.schildbach.wallet.rates.ExchangeRatesViewModel;
import de.schildbach.wallet.ui.dashpay.NotificationsActivity;
import de.schildbach.wallet.ui.dashpay.utils.ProfilePictureDisplay;
import de.schildbach.wallet_test.R;

@AndroidEntryPoint
public final class HeaderBalanceFragment extends Fragment implements SharedPreferences.OnSharedPreferenceChangeListener {

    private WalletApplication application;
    private AbstractBindServiceActivity activity;
    private Configuration config;

    private Boolean hideBalance;
    private View showBalanceButton;
    private TextView hideShowBalanceHint;
    private TextView caption;
    private View view;
    private CurrencyTextView viewBalanceDash;
    private CurrencyTextView viewBalanceLocal;
    private TextView notifications;
    private ImageButton notificationBell;
    private ImageView dashpayUserAvatar;
    private TextView syncingIndicator;

    private boolean showLocalBalance;

    private HeaderBalanceViewModel viewModel;
    private MainActivityViewModel mainActivityViewModel;
    private ExchangeRatesViewModel exchangeRatesViewModel;

    @Nullable
    private ExchangeRate exchangeRate = null;

    @Override
    public void onAttach(final Activity activity) {
        super.onAttach(activity);

        this.activity = (AbstractBindServiceActivity) activity;
        this.application = (WalletApplication) activity.getApplication();
        this.config = application.getConfiguration();
        hideBalance = config.getHideBalance();

        showLocalBalance = getResources().getBoolean(R.bool.show_local_balance);
    }

    @Override
    public void onActivityCreated(@androidx.annotation.Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initViewModel();
        setNotificationCount();
    }

    private void initViewModel() {
        exchangeRatesViewModel = new ViewModelProvider(this).get(ExchangeRatesViewModel.class);

        mainActivityViewModel = new ViewModelProvider(requireActivity()).get(MainActivityViewModel.class);
        mainActivityViewModel.getBlockchainStateData().observe(getViewLifecycleOwner(), blockchainState -> {
            viewModel.setBlockchainState(blockchainState);
            updateView();
        });
        mainActivityViewModel.getDashPayProfileData().observe(getViewLifecycleOwner(), dashPayProfile -> {
            updateView();
            setNotificationCount();
        });

        viewModel = new ViewModelProvider(this).get(HeaderBalanceViewModel.class);
        viewModel.getWalletBalanceData().observe(getViewLifecycleOwner(), balance -> {
            updateView();
        });
        viewModel.getNotificationCountData().observe(getViewLifecycleOwner(), notificationCount -> {
            setNotificationCount();
        });
        // don't query alerts if notifications are disabled
        if (config.areNotificationsDisabled()) {
            long lastSeenNotification = config.getLastSeenNotificationTime();
            AppDatabase.getAppDatabase().userAlertDaoAsync()
                    .load(lastSeenNotification).observe(getViewLifecycleOwner(), userAlert -> {
                if (userAlert != null) {
                    viewModel.forceUpdateNotificationCount();
                }
            });
        }
    }

    @Override
    public View onCreateView(final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {
        return inflater.inflate(R.layout.header_balance_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull final View view, final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        caption = view.findViewById(R.id.caption);
        hideShowBalanceHint = view.findViewById(R.id.hide_show_balance_hint);
        this.view = view;
        showBalanceButton = view.findViewById(R.id.show_balance_button);
        syncingIndicator = view.findViewById(R.id.balance_syncing_indicator);

        viewBalanceDash = view.findViewById(R.id.wallet_balance_dash);
        viewBalanceDash.setApplyMarkup(false);
        viewBalanceDash.setFormat(config.getFormat().noCode());
        viewBalanceDash.setAmount(Coin.ZERO);

        viewBalanceLocal = view.findViewById(R.id.wallet_balance_local);
        viewBalanceLocal.setInsignificantRelativeSize(1);

        view.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                hideBalance = !hideBalance;
                updateView();
            }
        });

        notifications = view.findViewById(R.id.notifications);
        notificationBell = view.findViewById(R.id.notification_bell);
        dashpayUserAvatar = view.findViewById(R.id.dashpay_user_avatar);

        notifications.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(NotificationsActivity.createIntent(getContext(), NotificationsActivity.MODE_NOTIFICATIONS));
            }
        });

        notificationBell.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(NotificationsActivity.createIntent(requireContext(), NotificationsActivity.MODE_NOTIFICATIONS));
            }
        });

        dashpayUserAvatar.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(NotificationsActivity.createIntent(requireContext(), NotificationsActivity.MODE_NOTIFICATIONS));
            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();

        exchangeRatesViewModel.getRate(config.getExchangeCurrencyCode()).observe(this,
                new Observer<ExchangeRate>() {
                    @Override
                    public void onChanged(ExchangeRate rate) {
                        if (rate != null) {
                            exchangeRate = rate;
                            updateView();
                        }
                    }
                });

        if (mainActivityViewModel.getHasIdentity()) {
            viewModel.forceUpdateNotificationCount();
        }
        if (config.getHideBalance()) {
            hideBalance = true;
        }
        config.registerOnSharedPreferenceChangeListener(this);

        updateView();
    }

    @Override
    public void onPause() {
        config.unregisterOnSharedPreferenceChangeListener(this);
        super.onPause();
    }

    private void setNotificationCount() {
        int notificationCount = viewModel.getNotificationCount();
        if (!mainActivityViewModel.getHasIdentity()) {
            notifications.setVisibility(View.GONE);
            notificationBell.setVisibility(View.GONE);
        } else if (notificationCount > 0) {
            notifications.setText(String.valueOf(notificationCount));
            notifications.setVisibility(View.VISIBLE);
            notificationBell.setVisibility(View.GONE);
        } else if (notificationCount == 0) {
            notifications.setVisibility(View.GONE);
            notificationBell.setVisibility(View.VISIBLE);
        }
    }

    private void updateView() {
        View balances = view.findViewById(R.id.balances_layout);
        DashPayProfile dashPayProfile = mainActivityViewModel.getDashPayProfile();
        ProfilePictureDisplay.display(dashpayUserAvatar, dashPayProfile, true);

        if (hideBalance) {
            caption.setText(R.string.home_balance_hidden);
            hideShowBalanceHint.setText(R.string.home_balance_show_hint);
            balances.setVisibility(View.INVISIBLE);
            showBalanceButton.setVisibility(View.VISIBLE);
            return;
        }
        balances.setVisibility(View.VISIBLE);
        caption.setText(R.string.home_available_balance);
        hideShowBalanceHint.setText(R.string.home_balance_hide_hint);
        showBalanceButton.setVisibility(View.GONE);

        if (!isAdded()) {
            return;
        }

        BlockchainState blockchainState = mainActivityViewModel.getBlockchainState();
        if (blockchainState != null && !blockchainState.isSynced()) {
            syncingIndicator.setVisibility(View.VISIBLE);
            startSyncingIndicatorAnimation();
        } else {
            syncingIndicator.setVisibility(View.GONE);
            if (syncingIndicator.getAnimation() != null) {
                syncingIndicator.getAnimation().cancel();
            }
        }

        if (!showLocalBalance) {
            viewBalanceLocal.setVisibility(View.GONE);
        }

        Coin balance = viewModel.getWalletBalance();
        if (balance != null) {
            viewBalanceDash.setAmount(balance);

            if (showLocalBalance) {
                if (exchangeRate != null) {
                    org.bitcoinj.utils.ExchangeRate rate = new org.bitcoinj.utils.ExchangeRate(Coin.COIN,
                            exchangeRate.getFiat());
                    final Fiat localValue = rate.coinToFiat(balance);
                    viewBalanceLocal.setVisibility(View.VISIBLE);
                    String currencySymbol = GenericUtils.currencySymbol(exchangeRate.getCurrencyCode());
                    viewBalanceLocal.setFormat(Constants.LOCAL_FORMAT.code(0, currencySymbol));
                    viewBalanceLocal.setAmount(localValue);
                } else {
                    viewBalanceLocal.setVisibility(View.INVISIBLE);
                }
            }
        } else {
            viewBalanceDash.setAmount(Coin.ZERO);
        }

        activity.invalidateOptionsMenu();
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (Configuration.PREFS_LAST_SEEN_NOTIFICATION_TIME.equals(key)) {
            viewModel.forceUpdateNotificationCount();
        }
    }

    private void startSyncingIndicatorAnimation() {
        Animation currentAnimation = syncingIndicator.getAnimation();
        if (currentAnimation == null || currentAnimation.hasEnded()) {
            AlphaAnimation alphaAnimation = new AlphaAnimation(0.2f, 0.8f);
            alphaAnimation.setDuration(833);
            alphaAnimation.setRepeatCount(Animation.INFINITE);
            alphaAnimation.setRepeatMode(Animation.REVERSE);
            syncingIndicator.startAnimation(alphaAnimation);
        }
    }

    private void showExchangeRatesActivity() {
        Intent intent = new Intent(getActivity(), ExchangeRatesActivity.class);
        getActivity().startActivity(intent);
    }
}

File: wallet/src/de/schildbach/wallet/ui/main/HeaderBalanceFragment.kt
/*
 * Copyright 2023 Dash Core Group.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.schildbach.wallet.ui.main

import android.os.Bundle
import android.view.View
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.launch
import de.schildbach.wallet.Constants
import de.schildbach.wallet_test.R
import de.schildbach.wallet_test.databinding.HeaderBalanceFragmentBinding
import org.bitcoinj.core.Coin
import org.dash.wallet.common.ui.viewBinding
import org.dash.wallet.common.util.GenericUtils
import org.dash.wallet.common.util.observe
import de.schildbach.wallet.util.toDashjFiat
import de.schildbach.wallet.util.setAmount

class HeaderBalanceFragment : Fragment(R.layout.header_balance_fragment) {
    private val viewModel by activityViewModels<MainViewModel>()
    private val binding by viewBinding(HeaderBalanceFragmentBinding::bind)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.walletBalanceDash.setApplyMarkup(false)
        binding.walletBalanceDash.setFormat(viewModel.balanceDashFormat)
        binding.walletBalanceDash.setAmount(Coin.ZERO)

        binding.walletBalanceLocal.setInsignificantRelativeSize(1f)
        binding.walletBalanceLocal.setStrikeThru(!Constants.IS_PROD_BUILD)
        requireView().setOnClickListener { viewModel.triggerHideBalance() }

        viewModel.exchangeRate.observe(viewLifecycleOwner) { updateBalance() }
        viewModel.totalBalance.observe(viewLifecycleOwner) { updateBalance() }

        // Phase 5d: after a committed cutover the "syncing" indicator reads the
        // SDK L1 scan instead of dashj (viewModel.sdkOwnsL1); collect that source
        // too so the indicator clears from whichever engine owns L1 this launch.
        // Pre-cutover (every install today) this is identical to the dashj-only
        // rendering.
        viewModel.isBlockchainSynced.observe(viewLifecycleOwner) { updateSyncingIndicator() }
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.sdkL1Synced.collect { updateSyncingIndicator() }
        }

        viewModel.hideBalance.observe(viewLifecycleOwner) { hideBalance ->
            binding.balanceGroup.isInvisible = hideBalance
            binding.showBalanceButton.isInvisible = !hideBalance
        }

        viewModel.showTapToHideHint.observe(viewLifecycleOwner) { showHint ->
            binding.hideBalanceHintText.isVisible = showHint != false
        }
    }

    private fun updateSyncingIndicator() {
        val sdkOwnsL1 = viewModel.sdkOwnsL1.value
        val isSynced = if (sdkOwnsL1) {
            viewModel.sdkL1Synced.value
        } else {
            viewModel.isBlockchainSynced.value == true
        }

        if (isSynced) {
            binding.syncingIndicator.isInvisible = true
            binding.syncingIndicator.animation?.cancel()
        } else {
            binding.syncingIndicator.isInvisible = false
            startSyncingIndicatorAnimation()
        }
    }

    private fun updateBalance() {
        val balance = viewModel.totalBalance.value ?: Coin.ZERO
        binding.walletBalanceDash.setAmount(balance)
        viewModel.exchangeRate.value?.let { exchangeRate ->
            val rate = org.bitcoinj.utils.ExchangeRate(
                Coin.COIN,
                exchangeRate.fiat.toDashjFiat()
            )

            val localValue = rate.coinToFiat(balance)
            val currencySymbol = GenericUtils.currencySymbol(exchangeRate.currencyCode)
            binding.walletBalanceLocal.setFormat(viewModel.fiatFormat.code(0, currencySymbol))
            binding.walletBalanceLocal.setAmount(localValue)
        }
    }

    private fun startSyncingIndicatorAnimation() {
        val currentAnimation = binding.syncingIndicator.animation

        if (currentAnimation == null || currentAnimation.hasEnded()) {
            val alphaAnimation = AlphaAnimation(0.2f, 0.8f)
            alphaAnimation.duration = 833
            alphaAnimation.repeatCount = Animation.INFINITE
            alphaAnimation.repeatMode = Animation.REVERSE
            binding.syncingIndicator.startAnimation(alphaAnimation)
        }
    }
}